### PR TITLE
Add `json.filter` builtin

### PR DESCRIPTION
**Pull request by patrick-east**
The new builtin takes an object and list of paths to keep.

Reference: #1617
Signed-off-by: Patrick East <east.patrick@gmail.com>

### CODE REVIEW

File: docs/content/policy-reference.md [hunk 1]
```diff
@@ -66,6 +66,13 @@ complex types.
 | <span class="opa-keep-it-together">``output := intersection(set[set])``</span> | ``output`` is the intersection of the sets in the input set  |
 | <span class="opa-keep-it-together">``output := union(set[set])``</span> | ``output`` is the union of the sets in the input set  |
 
+### Objects
+| Built-in | Description |
+| -------- | ----------- |
+| <span class="opa-keep-it-together">``filtered := json.filter(object, paths)``</span> | ``filtered`` is the remaining data from ``object`` with only keys specified in ``paths`` which is an array or set of strings. For example: ``json.filter({"a": {"b": "x", "c": "y"},}, ["a/b"]`` will result in ``{"a": {"b": "x"}}``). |
+
+The ``json`` object paths may reference into array values by using index numbers. For example with the object ``{"a": ["x", "y", "z"]}`` the path `a[1]` references `y`
```
Comment (tsandall) [inline docs/content/policy-reference.md]: Nit: I think this was supposed to be `/a/1` instead of `a[1]` here.

File: ast/term.go [hunk 2]
```diff
@@ -1959,6 +1970,51 @@ func (obj *object) insert(k, v *Term) {
...
+				if err != nil {
+					return nil, err
+				}
+				values = append(values, NewTerm(filteredValue))
+			}
+		}
+		return values, nil
+	case Object:
+		values := NewObject()
+		err := v.Iter(func(k *Term, v *Term) error {
```
Comment (tsandall) [inline ast/term.go]: Nit: One simple optimization would be to make this iterate over the smaller of the two objects.

File: docs/content/policy-reference.md [hunk 3]
```diff
@@ -66,6 +66,13 @@ complex types.
 | <span class="opa-keep-it-together">``output := intersection(set[set])``</span> | ``output`` is the intersection of the sets in the input set  |
 | <span class="opa-keep-it-together">``output := union(set[set])``</span> | ``output`` is the union of the sets in the input set  |
 
+### Objects
+| Built-in | Description |
+| -------- | ----------- |
+| <span class="opa-keep-it-together">``filtered := json.filter(object, paths)``</span> | ``filtered`` is the remaining data from ``object`` with only keys specified in ``paths`` which is an array or set of strings. For example: ``json.filter({"a": {"b": "x", "c": "y"},}, ["a/b"]`` will result in ``{"a": {"b": "x"}}``). |
```
[the remark this answers is not in the document]
Comment (patrick-east) [reply docs/content/policy-reference.md]: Done. (this was fixed over in #1939, I just squashed it into the wrong commit 😢 )

File: topdown/json.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+	// Build a list of filter strings
+	var filters []ast.String
+
+	switch v := operands[1].Value.(type) {
+	case ast.Array:
+		for _, f := range v {
+			filterString, ok := f.Value.(ast.String)
+			if !ok {
+				return builtins.NewOperandErr(2, "Expected a set or array containing only strings")
```
[the remark this answers is not in the document]
Comment (patrick-east) [reply topdown/json.go]: Done.

File: topdown/json.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+				if t.Value.Compare(ast.Null{}) == 0 {
+					break // done with path, another shorter path has already been here
+				} else if len(s) == 0 {
+					o.Insert(key, ast.NullTerm())
+				} else {
+					o = t.Value.(ast.Object)
+				}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (patrick-east) [reply topdown/json.go]: LGTM, latest version uses that one.

File: docs/content/policy-reference.md (hunk 1 above)
Comment (patrick-east) [reply docs/content/policy-reference.md]: Done.

File: ast/term.go (hunk 2 above)
Comment (patrick-east) [reply ast/term.go]: Done.

File: ast/term.go [hunk 6]
```diff
@@ -1959,6 +1970,51 @@ func (obj *object) insert(k, v *Term) {
 	obj.ground = obj.ground && k.IsGround() && v.IsGround()
 }
 
+func filterObject(o Value, filter Value) (Value, error) {
```
[the remark this answers is not in the document]
Comment (patrick-east) [reply ast/term.go]: Done.

**Comment by patrick-east** [comment]
Latest version supports arrays of path segments in addition to string paths

**Comment by patrick-east** [comment]
Latest update corrects behavior when the object contains sets. It now allows for indexing into them with the filters.

**Pull request merged**